Common/Algorithm/Framework/Alphas/Analysis/AlphaManager.cs

```diff
@@ -70,51 +65,94 @@ public class AlphaManager
 
-            _openAlphaContexts = new ConcurrentDictionary<Guid, AlphaAnalysisContext>();
+            _openAlphaContexts = new HashSet<AlphaAnalysisContext>();
+            _updatedAlphaContexts = new HashSet<AlphaAnalysisContext>();
             _closedAlphaContexts = new ConcurrentDictionary<Guid, AlphaAnalysisContext>();
-            _updatedAlphaContextsByAlphaId = new ConcurrentDictionary<Guid, AlphaAnalysisContext>();
         }
 
         /// <summary>
-        /// Add alphas to the manager from the collection
+        /// Add an extension to this manager
+        /// </summary>
+        /// <param name="extension">The extension to be added</param>
+        public void AddExtension(IAlphaManagerExtension extension)
+        {
+            _extensions.Add(extension);
+        }
+
+        /// <summary>
+        /// Initializes any extensions for the specified backtesting range
+        /// </summary>
+        /// <param name="start">The start date of the backtest (current time in live mode)</param>
+        /// <param name="end">The end date of the backtest (<see cref="Time.EndOfTime"/> in live mode)</param>
+        /// <param name="current">The algorithm's current utc time</param>
+        public void InitializeExtensionsForRange(DateTime start, DateTime end, DateTime current)
+        {
+            foreach (var extension in _extensions)
+            {
+                extension.InitializeForRange(start, end, current);
+            }
+        }
+
+        /// <summary>
+        /// Steps the manager forward in time, accepting new state information and potentialy newly generated alphas
         /// </summary>
-        /// <param name="collection">The alpha collection emitted from <see cref="IAlgorithm.AlphasGenerated"/></param>
-        public void AddAlphas(AlphaCollection collection)
+        /// <param name="frontierTimeUtc">The frontier time of the alpha analysis</param>
+        /// <param name="securityValuesCollection">Snap shot of the securities at the frontier time</param>
+        /// <param name="generatedAlphas">Any alpha generated by the algorithm at the frontier time</param>
+        public void Step(DateTime frontierTimeUtc, ReadOnlySecurityValuesCollection securityValuesCollection, AlphaCollection generatedAlphas)
         {
-            foreach (var alpha in collection.Alphas)
+            if (generatedAlphas != null && generatedAlphas.Alphas.Count > 0)
             {
-                // save initial security values and deterine analysis period
-                var initialValues = _securityValuesProvider.GetValues(alpha.Symbol);
-                var analysisPeriod = alpha.Period + TimeSpan.FromTicks((long) (_extraAnalysisPeriodRatio * alpha.Period.Ticks));
+                foreach (var alpha in generatedAlphas.Alphas)
+                {
+                    // save initial security values and deterine analysis period
+                    var initialValues = securityValuesCollection[alpha.Symbol];
```

Comment:
Possible key not found?

Reply:
hmm, good point, I suppose it's *possible* if the algorithm creates a symbol it doesn't have data for